freedomofpress/securedrop-client
# Delete conversation files and records as soon as server confirms scheduled deletion

# Description

Delete local files and relevant source records immediately as soon as conversation is scheduled for deletion on the server (do not wait for sync to delete).

Fixes #1429
Replaces https://github.com/freedomofpress/securedrop-client/pull/1424
Will be relevant for #1344

**High-level approach:**
- When the server has successfully scheduled files for deletion, files and records are deleted locally, and the source's UUID (not id) is inserted into a `deletedconversation` table (which is basically a flag). (We avoid storing source id and other PII/attributes.)
- For the first sync after the deletion request is made, if a source's UUID is in the `deletedconversation` table, we do not update their record. This is to avoid a situation where a stale sync re-populates source records that have been deleted (network race condition). (Note: a sync happens every 15 seconds)
- At the end of the first sync, the source's UUID is dropped from the `deletedconversation` table, and their records will be updated as normal for all subsequent syncs.
- This approach will also work for user deletion and avoiding stale syncs repopulating a deleted user's records. 

# Test Plan

**Test local deletion: happy path**
- Log into the client and select a source with replies, messages, and files. Download some files and observe them in a subdirectory of `~/.securedrop_client/data/`
- Safe-delete ("delete files and messages") for the source.
- [ ] Inspect the `~/.securedrop_client/data/` directory; the source's subdirectory should be removed and all relevant files deleted.  
- [ ] Inspect the `sources` table and note the source's id (not UUID). Query the `messages`, `replies`, `draftreplies`, and `files` tables for rows with the source `id` and ensure that no matching rows are present.

**Test local deletion with network failures**
- Log into the client and select a source with replies, messages, and files. Download some files and observe them in a subdirectory of `~/.securedrop_client/data/`
- Open the Qubes Settings -> network settings for sd-whonix, or open a terminal in sd-whonix.
- Safe-delete ("delete files and messages") for the source, then as soon as the UI confirms deletion, _quickly_ kill the network connection by setting the nevtm to `none` or issuing `sudo service tor stop` in the sd-whonix terminal.
- [ ] Inspect the `~/.securedrop_client/data/` directory; the source's subdirectory should be removed and all relevant files deleted.  
- [ ] Inspect the `sources` table and note the source's id and UUID. Query the `messages`, `replies`, `draftreplies`, and `files` tables for rows with the source `id` and ensure that no matching rows are present.
- [ ] Inspect the deletedconversation table and ensure that a row with the source `uuid` is present. 
- Re-connect the network and wait for sync.
- [ ] Again, query the `messages`, `replies`, `draftreplies`, and `files` tables for rows with the source `id` and ensure that no matching rows are present.
- [ ] Ensure the entry with source UUID has been removed from the deletedconversation table.

**Ensure no regression when deletion is done on the server**
- Download files for a source with files and messages.
- Delete the source's files and messages via the Journalist Interface.
- Wait for sync. 
- [ ] Ensure files and messages have been deleted from the client as in the steps above.

**Test upgrade of existing client**
- [ ] Ensure database upgrade works and local deletion works as above  

**(Added) No ghost replies** 
- Safe delete a conversation, then send reply A, then safe delete again.
- Send reply B
- [ ] Ensure reply A is not temporarily visible in the GUI (note: this is a GUI bug; it's not added to the database)

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:


If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:


If these changes modify the database schema, you should include a database migration. Please check as applicable:
